### DISCUSSION TITLE
Low-level Pascal translation/symbology

### DISCUSSION THREAD

**Discussion by texdraft** [shreevatsa/webWEB#14] [Ideas]
I am currently implementing Phase 3 in [my system](https://github.com/texdraft/web-to-html), and I'd like to discuss the matter of how non-identifier non-reserved-word (other than `and`, `in`, `or`, and `not`) tokens should appear in the output. (Of course this will be configurable to some extent.)

There are generally two ways to go: (1) copying `WEAVE` or (2) using Unicode characters with the proper semantics. In many cases we get the same result:

Pascal | Unicode/`WEAVE`
------ | -------
`(` | `(` U+0028 LEFT PARENTHESIS
`)` | `)` U+0029 RIGHT PARENTHESIS
`+` | `+` U+002B PLUS SIGN
`,` | `,` U+002C COMMA
`.` | `.` U+002E FULL STOP
`:` | `:` U+003A COLON
`;` | `;` U+003B SEMICOLON
`<` | `<` U+003C LESS-THAN SIGN
`=` | `=` U+003D EQUALS SIGN
`>` | `>` U+003E GREATER-THAN SIGN
`[` | `[` U+005B LEFT SQUARE BRACKET
`]` | `]` U+005D RIGHT SQUARE BRACKET
`not` | `¬` U+00AC NOT SIGN
`^` | `↑` U+2191 UPWARDS ARROW
`in` | `∈` U+2208 ELEMENT OF
`-` | `−` U+2212 MINUS SIGN
`and` | `∧` U+2227 LOGICAL AND
`or` | `∨` U+2228 LOGICAL OR
`<>` | `≠` U+2260 NOT EQUAL TO
`==` | `≡` U+2261 IDENTICAL TO
`<=` | `≤` U+2264 LESS-THAN OR EQUAL TO
`>=` | `≥` U+2265 GREATER-THAN OR EQUAL TO

Here are the others.

- `:=`, the “becomes” symbol. `WEAVE`'s translation can be rendered using `←` U+2190&nbsp;LEFTWARDS ARROW. Unicode also has `≔` U+2254&nbsp;COLON EQUALS. Even though I much prefer the look of `←`, it would arguably be more correct to use `≔`. For one, it more closely matches the appearance of `:=`. And even Algol 60 and Algol 68, which otherwise had extended character sets, went with “:=” for assignment (called “assignation” in Algol 68). However, `←` is also the notation Knuth uses in TAOCP.
- `/`,  the symbol denoting division with a `real` result. We have two contenders: `∕` U+2215&nbsp;DIVISION SLASH and `÷` U+00F7&nbsp;DIVISION SIGN. Both are semantically sound. `WEAVE` uses the former symbology.
- `*`, the multiplication symbol. Unicode has `×` U+00D7&nbsp;MULTIPLICATION SIGN, but `WEAVE`'s output has `∗` U+2217&nbsp;ASTERISK OPERATOR. There is also `✕` U+2715&nbsp;MULTIPLICATION X.
- `..`, the subrange symbol. `WEAVE`'s output is the equivalent of &langle;`.` U+002E&nbsp;FULL STOP, ` ` U+2009&nbsp;THIN SPACE, U+002E&nbsp;FULL STOP&rangle;. Although we could use this, there are two other options: `…` U+2026&nbsp;HORIZONTAL ELLIPSIS, which is the symbol `..` is mimicking (`..` was almost certainly chosen to simplify writing a lexical analyzer for the language) and `‥` U+2025&nbsp;TWO DOT LEADER.

But of course punctuators aren't the full story. There are also numbers and strings.

- Pascal strings `'…'` are translated by `WEAVE` directly, except that a monospace font is used. Call me crazy, but I propose that the translation should be the string's contents (monospace) in between `‘` U+2018&nbsp;LEFT SINGLE QUOTATION MARK and `’` U+2019&nbsp;RIGHT SINGLE QUOTATION MARK; double `''`s in the string will be changed to single `'`s. This is, however, unconventional.
- `WEB` strings `"…"` are translated by `WEAVE` like Pascal strings with a different delimiter. And similarly my proposal is to use `“` and `”` to surround the string's contents, in which `""` will be changed to `"`.
- Decimal numbers will be left alone as digit strings.
- Octal numbers are translated by `WEAVE` in italic type, preceded by `´` U+00B4&nbsp;ACUTE ACCENT. This isn't a bad choice, but I suggest using italic type but following the digits by a subscript `8`, rendered either by `₈` U+2088&nbsp;SUBCRIPT EIGHT or by a `<sub>` element.
- Hexadecimal numbers are translated by `WEAVE` in typewriter type, preceded by `˝` U+02DD&nbsp;DOUBLE ACUTE ACCENT. As for octal numbers, I suggest using typewriter type followed by a subscript `16`, rendered either by &langle;`₁` U+2081&nbsp;SUBSCRIPT ONE, `₆` U+2086&nbsp;SUBSCRIPT SIX&rangle; or by a `<sub>` element.
- Exponential notation in floating-point numbers is translated by `WEAVE` into actual exponential notation, so that `1.0e-4` becomes `$1.0\cdot10^{-4}$`. This can be done easily by using Unicode superscripts or `<sup>` elements; the only question is how the exponent should be set off from the rest of the number. `WEAVE` uses the equivalent of `⋅` U+22C5&nbsp;DOT OPERATOR followed by `10`. The most proper Unicode rendering would probably be `⏨` U+23E8&nbsp;DECIMAL EXPONENT SYMBOL, [following Algol 60](https://retrocomputing.stackexchange.com/questions/16082/was-there-an-input-device-capable-of-entering-all-algol-60-symbols-with-correct/16083?noredirect=1#comment68433_16083), although font support isn't great.

**Comment by shreevatsa** [comment]
I don't have an immediate answer here but it's amazing to see such minute care lavished on this system; I look forward to it. :-) But a "meta" comment in the meantime: personally one of my main reasons for interest in the (C)WEB parser/HTML translator is to have the option of reading those programs while moving a bit away from the idiosyncrasies of WEAVE or Knuth or Pascal, to something more familiar to typical programmers. Roughly, if one were writing a blog post with pseudocode, and one wouldn't adopt a particular syntax or typesetting choice, then it should be possible to have output that doesn't those choices. (BTW I agree with you about the translation of Pascal and WEB strings—https://github.com/nadder/rstex/pull/2 is vaguely related—and also octal and hexadecimal numbers.) So it's not too crazy for me to imagine `:=` remaining `:=` or even `=`, or even for the entire output to be monospace with no non-ASCII characters at all. So this part above:

> (Of course this will be configurable to some extent.)

is the most exciting to me :-) 

(Maybe the output could have things like `<span class=assignment>:=</span>` and CSS could replace the contents….)

**Reply by texdraft** [reply]
Regarding pseudocode: It would be quite possible to modify/extend my Phase 2 to produce syntax trees, which might then be used to translate the Pascal to any language or notation. (Perhaps I'll write a Pascal&rarr;Lisp translator, and then we can have an implementation of TeX in Lisp.)

Regarding strings and numbers: I didn't mention in the main post that you'll be able to hover over strings to see their numeric value, and you can hover over integer literals to see their representation in other radices.

Regarding CSS for symbols: Unfortunately [CSS isn't the right tool](https://stackoverflow.com/questions/7896402/how-can-i-replace-text-with-css) for this. (Even if the hack at the linked question is adopted, pseudo-elements can't be interacted with; in particular they cannot be selected and [copied to the clipboard](https://github.com/shreevatsa/webWEB/discussions/14#discussioncomment-995411).) Javascript, on the other hand, could do it easily, although there might be a bit of a performance hit considering the massive number of tokens it will have to crawl through.

**Comment by dahosek** [comment]
I wonder though whether this sort of translation makes sense for an online presentation where it’s more likely that readers will do a direct copy and paste of code. After all. if I’m re-typing, it’s no big deal to write := when I see ←, bur it’s a pain if I have to go through and fix that after a copy-paste. I imagine that there might be some JavaScript trickery possible to undo the translations though.

-dh

**Reply by texdraft** [reply]
Copying and pasting is a good point. (See also [this issue](https://github.com/ascherer/cweb/issues/21) on the `CWEB` repository.) @shreevatsa's site has [some `WEB` programs uploaded](https://shreevatsa.net/tex/program#b2-smaller-other-programs), with buttons to toggle viewing the rendered SVG form vs. the raw `WEB` input vs. the TeX output of `WEAVE`&mdash;perhaps such a feature would be nice to have (probably allowing the toggle to be &ldquo;local&rdquo; to one section as well as global), and I don't think it would be hard for me to add.

I think my program retains enough information about the input to reconstruct Pascal source code (and this is a deliberate design feature of @zmajeed's `CWEB` reader), although all spacing/indentation is stripped away. However, Phase 2 goes back and adds in formatting details, which could be used in reproducing the un-translated form.

There is a similarity here with the issue of Algol 60 and Algol 68's &ldquo;[levels of representation](https://en.wikipedia.org/wiki/ALGOL_68#Program_representation)&rdquo;; the woven output is analogous to the &ldquo;representation language&rdquo;, while the Pascal input is a &ldquo;hardware representation&rdquo;. (In fact, I believe that the design of `WEB` was somewhat influenced by this property of Algol.)
